doc/decisions/creating_keysets.md

```diff
@@ -0,0 +1,95 @@
+# Creating KeySets
+
+## Problem
+
+The API that `libelektra-core` defines for creating KeySets must be possible to implement in other languages without complicated workarounds.
+In particular, another implementation should not need to resort to adding a small amount of C to "bridge the gap".
+
+## Constraints
+
+- There should be a way to create a `KeySet *` containing an arbitrary number of keys in a single expression.
+
+  This is very useful e.g. when initializing default configurations.
+
+- Ideally, there should be a way to create a `KeySet *` directly from a `Key *[]`.
+
+  A `Key *[]` can be created in a single expression `((Key *[]){ key1, key2, key3 })` and then sorted with a single `qsort`.
+  Most other methods require repeated `ksAppendKey`, which essentially amounts to Insertion Sort instead of Quicksort.
+
+## Assumptions
+
+- Many languages have trouble with C's variadic arguments.
+- Languages will provide a secondary library with more idiomatic APIs on top of `libelektra-core`.
+  If necessary, this library will also use unstable APIs from `libelektra-core` and in particular access `struct _KeySet` directly.
+
+## Considered Alternatives
+
+- Provide `KeySet * ksFromArray (const Key ** keys, size_t size)` within `libelektra-core`.
+
+  The separate size argument that is required in C could be avoided in most other languages.
+  This API is also not entirely minimal.
+
+## Decision
+
+`libelektra-core` only provides the absolute minimal API to create a `KeySet *`:
+
+```c
+/**
+ * Allocates a new empty keyset, with space reserved for @p alloc keys.
+ */
+Key * ksNew (size_t alloc);
+```
+
+Other libraries will provide APIs on top of this function.
+These functions should be more idiomatic to their target language.
+
+### Rust
+
+For example, `libelektra-rust` (targeted at Rust) could provide:
+
+```rust
+trait KeySet {
+    fn new<I: ExactSizeIterator<Item = Key>>(keys: I) -> Self;
+}
+```
```

Comment:
@lawli3t please check that this Rust API makes sense (at least in theory, it doesn't have to be implemented exactly like this)

Reply:
Not sure about ExactSizeIterator. Theoretically you could lazily generate KeySets where you don't necessarily know the exact size.

Maybe we would need to separate this? Firstly we should implement (FromIterator)[https://doc.rust-lang.org/std/iter/trait.FromIterator.html] which provides lots of syntactic sugar for devs and is the common trait for Classes that can be instantiated from iterators. Additionally I would add

```rust
trait KeySet {
    fn new (keys: &[Key]) -> Self;
}
```

as this is way more liberal in what it accepts.

Although a certain case could be made that we split this up even further into something like this, if we follow [1] strictly

```rust
trait KeySet {
    // default constructor, only default settings (Default trait)
    fn new () -> Self;
    // for specifying the size
    fn with_capacity(size: usize) -> Self;
    // for creating from a collection / list &[Key] because you can then supply a multitude of things (slices, vectors,...)
    fn from_slice(keys: &[Key]) -> Self;
}

impl FromIterator<Key> for KeySet {
    fn from_iter<I: IntoIterator<Item=Key>>(iter: I) -> Self {
        // ....
    }
}
```

We should also implement `IntoIterator` then, as this would give us a free copy constructor and looping and other stuff....

of course that would make it a bit tricky to create a keySet with a specific size and import Keys in one statement, but from what i've seen from other rust libraries this is the general idiom and no problem at all. You just create the KeySet first and then add all the Keys. It also shouldn't matter performancewise.

[1] https://deterministic.space/elegant-apis-in-rust.html#techniques

Reply:
Okay, then I'll change the example to simply `new (keys: &[Key])` and you just implement whatever you think is best.